src/libmongoc/src/mongoc/mongoc-async-cmd.c

```diff
@@ -134,42 +134,47 @@ mongoc_async_cmd_run (mongoc_async_cmd_t *acmd)
    return false;
 }
 
-void
-_mongoc_async_cmd_init_send (const mongoc_opcode_t cmd_opcode,
+static void
+_mongoc_async_cmd_init_send (const int32_t cmd_opcode,
                              mongoc_async_cmd_t *acmd,
                              const char *dbname)
 {
-   acmd->rpc.header.msg_len = 0;
-   acmd->rpc.header.request_id = ++acmd->async->request_id;
-   acmd->rpc.header.response_to = 0;
+   BSON_ASSERT (cmd_opcode == MONGOC_OP_CODE_QUERY ||
+                cmd_opcode == MONGOC_OP_CODE_MSG);
 
-   if (MONGOC_OPCODE_QUERY == cmd_opcode) {
-      acmd->ns = bson_strdup_printf ("%s.$cmd", dbname);
-      acmd->rpc.header.opcode = MONGOC_OPCODE_QUERY;
-      acmd->rpc.query.flags = MONGOC_QUERY_SECONDARY_OK;
-      acmd->rpc.query.collection = acmd->ns;
-      acmd->rpc.query.skip = 0;
-      acmd->rpc.query.n_return = -1;
-      acmd->rpc.query.query = bson_get_data (&acmd->cmd);
-      acmd->rpc.query.fields = NULL;
-   }
+   int32_t message_length = 0;
 
-   if (MONGOC_OPCODE_MSG == cmd_opcode) {
-      acmd->rpc.header.opcode = MONGOC_OPCODE_MSG;
+   message_length += mcd_rpc_header_set_message_length (acmd->rpc, 0);
+   message_length +=
+      mcd_rpc_header_set_request_id (acmd->rpc, ++acmd->async->request_id);
+   message_length += mcd_rpc_header_set_response_to (acmd->rpc, 0);
+   message_length += mcd_rpc_header_set_op_code (acmd->rpc, cmd_opcode);
 
-      acmd->rpc.msg.msg_len = 0;
-      acmd->rpc.msg.flags = 0;
-      acmd->rpc.msg.n_sections = 1;
-      acmd->rpc.msg.sections[0].payload_type = 0;
-      acmd->rpc.msg.sections[0].payload.bson_document =
-         bson_get_data (&acmd->cmd);
+   if (cmd_opcode == MONGOC_OP_CODE_QUERY) {
+      acmd->ns = bson_strdup_printf ("%s.$cmd", dbname);
+      message_length += mcd_rpc_op_query_set_flags (
+         acmd->rpc, MONGOC_OP_QUERY_FLAG_SECONDARY_OK);
+      message_length +=
+         mcd_rpc_op_query_set_full_collection_name (acmd->rpc, acmd->ns);
+      message_length += mcd_rpc_op_query_set_number_to_skip (acmd->rpc, 0);
+      message_length += mcd_rpc_op_query_set_number_to_return (acmd->rpc, -1);
+      message_length +=
+         mcd_rpc_op_query_set_query (acmd->rpc, bson_get_data (&acmd->cmd));
+   } else {
+      mcd_rpc_op_msg_set_sections_count (acmd->rpc, 1u);
+      message_length +=
+         mcd_rpc_op_msg_set_flag_bits (acmd->rpc, MONGOC_OP_MSG_FLAG_NONE);
+      message_length += mcd_rpc_op_msg_section_set_kind (acmd->rpc, 0u, 0);
+      message_length += mcd_rpc_op_msg_section_set_body (
+         acmd->rpc, 0u, bson_get_data (&acmd->cmd));
    }
 
+   mcd_rpc_message_set_length (acmd->rpc, message_length);
+
    /* This will always be hello, which are not allowed to be compressed */
-   _mongoc_rpc_gather (&acmd->rpc, &acmd->array);
-   acmd->iovec = (mongoc_iovec_t *) acmd->array.data;
-   acmd->niovec = acmd->array.len;
-   _mongoc_rpc_swab_to_le (&acmd->rpc);
+   acmd->iovec = mcd_rpc_message_to_iovecs (acmd->rpc, &acmd->niovec);
```

Comment:
Aside: I seem to recall some confusion expressed about the meaning of "gather" in this context. For reference, the word refers to "scatter/gather I/O", which is just a different name for vectored I/O.

Reply:
Personally, I am constantly confused by the perspective of the terminology. I am tempted to interpret the operations as "gathering" data into an RPC object and "scattering" data into iovec data structures (where the perspective is relative to the RPC object). This happens be the exact opposite of how libmongoc currently defines them: `mongoc_rpc_gather` "gathers" data from an RPC object into an array of iovecs and (this is truly the most confusing part) `mongoc_rpc_scatter` "scatters" data from a buffer into an RPC object (???). It doesn't help that "scatter/gather I/O" doesn't seem to rigorously define these terms anyways (their concern is with the vectored read/write interfaces, not with how those vectors are created), so `mcd-rpc` deliberately does away with that terminology in favor of a much more concrete vocabulary (from data, to iovecs).